### ISSUE TITLE
crash on `each_with_index` usage

### ISSUE THREAD

**Issue by marcandre** [ruby/typeprof#14]
Is it too early for bug reports?

I tried `typeprof` on `lib/matrix.rb` and all the methods that call `Array#each_with_index` crash it with:

```
DEBUG: ../dev/lib/matrix.rb:620 (each_with_index) PC=68 insn=send sp=1
["../dev/lib/matrix.rb:620", "[error] undefined method: String#each_with_index"]
["../dev/lib/matrix.rb:620", "[error] undefined method: nil#each_with_index"]
bundler: failed to load command: typeprof (/Users/mal/.rvm/gems/ruby-head/bin/typeprof)
RuntimeError: nil
  /Users/mal/ruby/typeprof/lib/typeprof/type.rb:205:in `block in initialize'
```

If I comment out these methods, `typeprof` seems to take many minutes analysing a few (unusual) lines [like these](https://github.com/ruby/matrix/blob/master/lib/matrix.rb#L1312-L1324).

After ~12 minutes (I have a 2013 MacbookAir 😅 ), it crashes with:

```sh
NoMethodError: undefined method `each' for nil:NilClass
  /Users/mal/ruby/typeprof/lib/typeprof/analyzer.rb:1976:in `block (2 levels) in show_block_signature'
  /Users/mal/ruby/typeprof/lib/typeprof/type.rb:94:in `each_child_global'
  /Users/mal/ruby/typeprof/lib/typeprof/analyzer.rb:1966:in `block in show_block_signature'
  /Users/mal/ruby/typeprof/lib/typeprof/analyzer.rb:1965:in `each'
  /Users/mal/ruby/typeprof/lib/typeprof/analyzer.rb:1965:in `show_block_signature'
  /Users/mal/ruby/typeprof/lib/typeprof/type.rb:803:in `screen_name'
  /Users/mal/ruby/typeprof/lib/typeprof/analyzer.rb:2023:in `show_method_signature'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:139:in `block (3 levels) in conv_class'
  /Users/mal/ruby/typeprof/lib/typeprof/utils.rb:101:in `each_key'
  /Users/mal/ruby/typeprof/lib/typeprof/utils.rb:101:in `each'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:131:in `block (2 levels) in conv_class'
  /Users/mal/ruby/typeprof/lib/typeprof/utils.rb:101:in `each_key'
  /Users/mal/ruby/typeprof/lib/typeprof/utils.rb:101:in `each'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:125:in `block in conv_class'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:124:in `each'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:124:in `conv_class'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:241:in `block in build_class_hierarchy'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:239:in `each'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:239:in `map'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:239:in `build_class_hierarchy'
  /Users/mal/ruby/typeprof/lib/typeprof/export.rb:214:in `show'
  /Users/mal/ruby/typeprof/lib/typeprof/analyzer.rb:857:in `report'
  /Users/mal/ruby/typeprof/lib/typeprof/config.rb:90:in `analyze'
  /Users/mal/ruby/typeprof/exe/typeprof:6:in `<top (required)>'
  /Users/mal/.rvm/gems/ruby-head/bin/typeprof:23:in `load'
  /Users/mal/.rvm/gems/ruby-head/bin/typeprof:23:in `<top (required)>'
```

**Comment by mame** [comment]
> Is it too early for bug reports?

No, I really appreciate it!

> all the methods that call Array#each_with_index crash

I could reproduce the crash. I have an idea why it fails (it reaches an inconsistent state), but I need to take time to investigate how it reaches the inconsistent state. I will check and fix it.
(Detail: It looks that an wrong-arity generic instance is created. Maybe, RBS declares a two-type-variable class like `class Foo[X, Y]`, but one-type-variable class instance like `Foo[X]` seems to be created for a unknown reason.)

> After ~12 minutes (I have a 2013 MacbookAir 😅 ), it crashes with:

I didn't reproduce it, but I think I fixed the issue at 955df724aebe4035a00b9e6619afa95b3105269b.

Sorry for ~12 minutes 😭. (Good news: the crash occurs when printing RBS, so the analysis itself is already finished. It will take no more time than 12 minutes 😄)
I know that the analysis performance is a serious problem of typeprof. Though it is fundamentally a slow approach, there are much room to improve the performance. But now I prioritize the not-implemented-yet features.

Thank you!

**Comment by marcandre** [comment]
> > After ~12 minutes (I have a 2013 MacbookAir 😅 ), it crashes with:
> 
> I didn't reproduce it, but I think I fixed the issue at [955df72](https://github.com/ruby/typeprof/commit/955df724aebe4035a00b9e6619afa95b3105269b).

I do get an analysis now, great. 🎉 

> Sorry for ~12 minutes 😭. (Good news: the crash occurs when printing RBS, so the analysis itself is already finished. It will take no more time than 12 minutes 😄)
>
> I know that the analysis performance is a serious problem of typeprof. Though it is fundamentally a slow approach, there are much room to improve the performance. But now I prioritize the not-implemented-yet features.

Yes, of course. Also, I just realized that I was out of Ram (I need a new machine), and Ruby was taking 2G and swapping, so probably much better on a good machine.

Now the results are not particularly usable...
<details>
<summary>This is what I get for `Matrix#initialize`</summary>

```
def initialize : ((Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Complex | Float | Integer | Matrix | Rational | String | Vector] |
 Array[Complex | Float | Integer | Rational] |
 Array[Integer] |
 Array[Integer] |
 Array[String?] |
 Array[Vector] |
 Array[untyped] |
 Complex | Enumerator[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Complex | Float | Integer | Matrix | Rational | String | Vector] |
 Array[Integer] |
 Array[String?] |
 Array[Vector] |
 Array[untyped] |
 Integer | String)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Complex | Float | Integer | Matrix | Rational | String | Vector] |
 Array[Integer] |
 Array[String?] |
 Array[untyped] |
 Integer | String)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Integer] |
 Array[String?] |
 Array[untyped] |
 Integer | String)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Integer] |
 Array[String?] |
 Integer | String)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Integer] |
 Array[String?] |
 Integer | String)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[Complex | Float | Integer | Matrix | Rational | String | Vector] |
 Array[Vector] |
 Complex | Float | Integer | Matrix | Rational | String | Vector)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[Vector] |
 Array[Vector] |
 Complex | Enumerator[untyped] |
 Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[Vector] |
 Complex | Enumerator[untyped] |
 Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Complex | Enumerator[untyped] |
 Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Array[(Complex | Float | Integer | Numeric | Rational | String | Vector)?] |
 Complex | Float | Integer | Numeric | Rational | String | Vector)?] |
 Array[(Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Integer] |
 Array[String?] |
 Array[Vector?] |
 Integer | String)?] |
 Array[(Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Integer] |
 Array[String?] |
 Array[nil] |
 Integer | String)?] |
 Array[(Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Integer] |
 Array[String?] |
 Array[nil] |
 Integer | String)?] |
 Array[(Array[(Complex | Float | Integer | Rational | String)?] |
 Array[(Integer | String)?] |
 Array[Integer] |
 Array[String?] |
 Integer | String)?] |
 Array[(Array[(Complex | Float | Integer | Rational | String)?] |
 Complex | Float | Integer | Rational | String)?] |
 Array[(Array[(Integer | String)?] |
 Array[Integer] |
 Array[String?] |
 Integer | String)?] |
 Array[(Array[(Integer | String)?] |
 Array[Integer] |
 Integer)?] |
 Array[(Array[Complex | Float | Integer | Matrix | Rational | String | Vector] |
 Complex | Float | Integer | Matrix | Rational | String | Vector)?] |
 Array[(Array[String?] |
 Integer | String)?] |
 Array[(Array[String?] |
 String)?] |
 Array[(Array[Vector] |
 Array[Vector] |
 Complex | Enumerator[untyped] |
 Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Array[Vector] |
 Array[Vector] |
 Complex | Enumerator[untyped] |
 Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Array[Vector] |
 Complex | Enumerator[untyped] |
 Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Array[Vector] |
 Complex | Float | Integer | Matrix | Rational | String | Vector)?] |
 Array[(Complex | Enumerator[untyped] |
 Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Matrix | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Matrix | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String | Vector)?] |
 Array[(Complex | Float | Integer | Numeric | Rational | String)?] |
 Array[(Complex | Float | Integer | Rational | String)?] |
 Array[Array[Integer]] |
 Array[Array[untyped]?] |
 Array[Array[untyped]] |
 Array[Integer?] |
 Array[String?] |
 Array[Vector?] |
 Array[nil] |
 String)?, ?(Complex | Float | Integer | Matrix | Range | Rational)?) -> ((Complex | Float | Integer | Matrix | Range | Rational)?)
```
</details>

First thing I notice is that there are a lot of optionals, which come from a series of deductions of methods that it thinks could return `nil`: `convert_to_array` <= `Vector#to_a` <= `Vector@elements` <= ?

I will feed it some basic information and see what happens.

A clear bug is that some elements seem repeated... ` Array[Vector] | Array[Vector] | ...` should I open a different issue?

**Comment by mame** [comment]
> Array[Vector] | Array[Vector] | ... should I open a different issue?

Yep, please. I have no idea why such a type is produced, so will see it later. Thanks!

**Comment by mame** [comment]
I think I finally fixed the first issue. The fix has only three lines, but it took a day 😫

Odd result like `Array[Vector] | Array[Vector]` is not fixed yet (I'll address it in #16), but anyway, now `bundle exec exe/typeprof ../ruby/lib/matrix.rb` runs in about 40 seconds on my machine without abort.

**Comment by marcandre** [comment]
>  runs in about 40 seconds on my machine without abort.

OMG, what a difference. Ram usage went from 2G+ to ~200MB too 💪
